File: 02_equacoes-alg-trans/c/birge_vieta.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main(void) {
    double erro = pow(10, -8);
    double coeficientes[] = {1, 0, 2, -1};
    double degree = 3;
    int conta_ciclos = 0;

    double b = coeficientes[0], c = b;
    double xk = 1, r = 1;

    while (fabs(r) > erro) {
        conta_ciclos++;
        double b = coeficientes[0], c = b;
        for (int i = 1; i < degree + 1; i++) {
            b = b * xk + coeficientes[i];
            if (i < degree)
                c = c * xk + b;
        }

        xk = xk - b / c;
        r = b;

        printf("R = %.16f\n", b);
        printf("R1 = %.16f\n", c);
        printf("xk = %.16f\n", xk);
    }

    printf("Operação finalizada!\n");
    printf("xk = %.16f\n", xk);
    printf("p(xk) = %.16f\n", r);
    printf("Ciclos = %d", conta_ciclos);
    return 0;
}

File: 02_equacoes-alg-trans/python/birge_vieta.py
erro = 10 ** (-8); conta_ciclos = 0
xk = 1 # estimativa inicial

coeficientes = [1, 0, 2, -1]
r = 1
while (abs(r) > erro):
    conta_ciclos += 1
    b = coeficientes[0]; c = b
    for i in range(1, len(coeficientes)):
        b = b * xk + coeficientes[i]
        if (i < len(coeficientes) - 1):
            c = c * xk + b;
    
    xk = xk - b / c
    r = b

    print(f'R = {b:.16f}')
    print(f'R1 = {c:.16f}')
    print(f'xk = {xk:.16f}')

print('Operaçao finalizada!\n')    
print(f'xk = {xk:.16f}')
print(f'p(xk) = {b:.16f}')
print(f'ciclos: {conta_ciclos}')

